finat/point_set.py

```diff
@@ -68,6 +68,43 @@ def expression(self):
         return gem.Literal(self.point)
 
 
+class UnknownPointSingleton(AbstractPointSet):
+    """A point set representing a single point with unknown location but known
+    `gem.Variable` expression.
+
+    Stored points are `numpy.nan`s to indicate that they are not known.
+
+    In the future this can be easily replaced with an equivalent for a whole
+    point set representing multiple points though storing the full array of
+    `numpy.nans` may become undesirable and require interface changes."""
+
+    def __init__(self, point_expr):
+        """Build a PointSingleton from a gem expression for a single point.
+
+        :arg point_expr: A `gem.Variable` expression representing a single
+            point of shape (D,) where D is the dimension of the point.
+            Since this represents a single point and is a `gem.Variable` it
+            should have no free indices."""
+        assert isinstance(point_expr, gem.Variable)
+        # Check we only have 1 point
+        assert point_expr.free_indices == ()
+        assert len(point_expr.shape) == 1
+        self.expression = point_expr
+        self.points = numpy.full(point_expr.shape, numpy.nan).reshape(1, -1)
```

Comment:
Idea from @wence- 
> If you need something that advertises the "meta-data" part of the numpy array just make a mock object.

Reply:
So I've tried adding this
```python
class UnknownPoints():
    """An object that behaves like a set of known points (with
    appropriate length and size) but which is unknown. For use with
    :class:`AbstractPointSet`s whos points are not known at compile
    time. If indexed, return numpy.nans."""
    def __init__(self, shape):
        """
        :arg shape: The shape of the unknown set of N points of shape
            (N, D) where D is the dimension of each point.
        """
        assert len(shape) == 2
        self.shape = shape

    def __len__(self):
        return self.shape[0]
```
the trouble is that `ps.points` needs to be iterable and subscriptable when passed to, for example, FIAT on [this line](https://github.com/FInAT/FInAT/blob/1e0e97ddce99cd3ec8f09169eea8f35980749b42/finat/quadrature_element.py#L93).
so I either need to add special cases (yuck) or do something like the below (but with correct indexing) 
```python
    def __getitem__(self, idx):
        return numpy.full(self.shape[1], numpy.nan)
```

Reply:
I guess this pretty much suggests that something is not right in the design. I think that one can't plausibly construct a fiat equivalent to these things. 

But just slap an array of nans in there for now.

Reply:
Okay I will leave as is for now. Have added a todo comment and will make an issue

Reply:
Created issue #83 .